hw/ip/otbn/util/shared/operand.py

```diff
@@ -309,47 +312,56 @@ def op_val_to_enc_val(self,
         if self.width is None:
             return None
 
+        # The operand can have two offsets applied to it. The PC offset is
+        # applied before shifting, the fixed_offset is applied after shifting.
         if self.pc_rel:
             # If the operand is PC-relative and we don't have cur_pc, give up.
             if cur_pc is None:
                 return None
 
             # Otherwise, we need to encode the offset.
-            offset_val = op_val - cur_pc
+            pc_offset_val = op_val - cur_pc
         else:
-            offset_val = op_val
+            pc_offset_val = op_val
 
-        # First, try to shift right. Check that we won't clobber any low bits.
+        # Following PC offset, try to shift right. Check that we won't clobber
+        # any low bits.
         shift_mask = (1 << self.shift) - 1
-        if offset_val & shift_mask:
+        if pc_offset_val & shift_mask:
             raise ValueError('Cannot encode the value {}: the operand has a '
                              'shift of {}, but that would clobber some bits '
                              '(because {} & {} = {}, not zero).'
-                             .format(offset_val, self.shift,
-                                     offset_val, shift_mask,
-                                     offset_val & shift_mask))
+                             .format(pc_offset_val, self.shift,
+                                     pc_offset_val, shift_mask,
+                                     pc_offset_val & shift_mask))
 
-        shifted = offset_val >> self.shift
+        shifted = pc_offset_val >> self.shift
+
+        # Finally apply the fixed offset
+        final_val = shifted - self.fixed_offset
 
         rng = self.get_rel_range()
         assert rng is not None
         lo, hi = rng
 
-        if not (lo <= offset_val <= hi):
-            shifted_msg = (', which shifts down to {}'.format(shifted)
+        # Range is given factoring in the fixed offset so check pre-fixed offset
+        # value against the range. final_val must already be computed so it can
+        # be used in the error message.
```

Comment:
I don't think I understand this. The point of the check below was to make sure the encoded value fits in the given number of bits. So I'd expect to be checking `final_val` here. Have I missed something?

Reply:
The issue is the `get_rel_range` function, which is also used to generate the documentation. The fixed offset gets applied to the low and high portions of this range (giving us the correct range of operands when generating the documentation). If you leave out the fixed offset from that you can drop this comment and the `pc_offset_val` below becomes `final_val` but then you don't have correct documentation.

My fix here is a little awkward, if we want this check to be a simple, is final value in range of allowable encoded values then we'll want a third range function `get_enc_range` or similar, or add a new `get_doc_range` for generating the documentation.

I should also alter the shifts down message to something like ', which encodes to ...' as it's the value with shift and offset applied and ensure it's displayed for either non-zero shift or non-zero offset.

Reply:
I favour making the "obvious path" in the code the one taken by the assembler and disassembler. If we have to do clever hoop-jumping, let's put that in the documentation generation code.

Reply:
Just realised this check isn't a simple is final value in range of allowable encoded values in the original code as it's done pre-shift, but uses the post-shift (i.e. final encoded) value in an error message. I think the best way forward for an "obvious path" perspective is:

1. Add a third `get_enc_range`, this gives the allowable range of values when encoded
2. This check gets its `lo` and `hi` from `get_enc_range` and checks them against `final_val`
3. Rename `get_rel_range` to `get_doc_range`
4. If there's an error in the final check use `get_doc_range` to get user-friendly `lo` and `hi` to generate the error message with.

Reply:
That sounds like a good idea.